### DISCUSSION TITLE
Typedef of swell-js sucks at highest level

### DISCUSSION THREAD

**Discussion by VojtaSim** [swellstores/community#232] [Ideas]
I haven't seen this concern raised anywhere so I feel like it needs to mentioned.
Working with the swell-js library in a TS project makes me cry...

1. the typedef doesn't correlate to what's actually received from the API (e.g. `attributes` on Product)
2. it doesn't account for excerpted entities, sometimes product isn't included in its whole form (e.g. Product in bundle product)
3. it doesn't differentiate between input types and result types - most of the properties are marked as optional `?` because of this.

swell-node could also use some typedefs... I've spent hours creating types according to docs just to be able to work with it.

Please do something about that 🙏

**Comment by maarten2424** [comment]
Hey @VojtaSim,
Have you considered using GraphQL in your project? Then you can make sure that your types are always up to date with your storefront.
You can set up GraphQL codegen to generate the correct types based on your GraphQL queries.

Here's an example of GraphQL codegen with `graphql-request`:

```codegen.ts
import { CodegenConfig } from "@graphql-codegen/cli";
import swellConfig from "./swell.config";

const config: CodegenConfig = {
  schema: {
    [swellConfig.url]: {
      headers: {
        Authorization: swellConfig.token,
      },
    },
  },
  documents: ["src/**/*.gql"],
  generates: {
    "./src/__generated__/sdk.ts": {
      plugins: [
        "typescript",
        "typescript-operations",
        "typescript-graphql-request",
      ],
      config: {
        rawRequest: true,
      },
    },
  },
};

export default config;

```

**Reply by VojtaSim** [reply]
I'm not really inclined towards GraphQL as we're using TRPC on the project.
